cirq-core/cirq/transformers/gauge_compiling/gauge_compiling.py

```diff
@@ -0,0 +1,183 @@
+# Copyright 2024 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Creates the abstraction for gauge compiling as a cirq transformer."""
+
+from typing import Callable, Tuple, Optional, Sequence, Union, List
+import abc
+import itertools
+import functools
+from dataclasses import dataclass
+
+import numpy as np
+
+from cirq.transformers import transformer_api
+from cirq import ops, circuits
+
+
+_SINGLE_QUBIT_GATES_T = Optional[Union[ops.Gate, Sequence[ops.Gate]]]
+
+
+class Gauge(abc.ABC):
+    """A gauge replaces a two qubit gate with an equivalent subcircuit.
+    0: pre_q0───────two_qubit_gate───────post_q0
+                        |
+    1: pre_q1───────two_qubit_gate───────post_q1
+
+    The Gauge class in general represents a family of closely related gauges
+    (e.g. random z-rotations); Use `sample` method to get a specific gauge.
+    """
+
+    def weight(self) -> float:
+        """Returns the relative frequency for selecting this gauge."""
+        return 1.0
+
+    @abc.abstractmethod
+    def sample(self, gate: ops.Gate, prng: np.random.Generator) -> "ConstantGauge":
+        """Returns a ConstantGauge sampled from a family of gauges.
+
+        Args:
+            gate: The two qubit gate to replace.
+            prng: A numpy random number generator.
+
+        Returns:
+            A ConstantGauge.
+        """
+
+
+@dataclass(frozen=True)
+class ConstantGauge(Gauge):
+    """A gauge that replaces a two qubit gate with a constant gauge."""
+
+    two_qubit_gate: ops.Gate
+    pre_q0: _SINGLE_QUBIT_GATES_T = None
+    pre_q1: _SINGLE_QUBIT_GATES_T = None
+    post_q0: _SINGLE_QUBIT_GATES_T = None
+    post_q1: _SINGLE_QUBIT_GATES_T = None
```

Comment:
For the sake of type simplicity - can this be just a `tuple[ops.Gate, ...]` with a `()` default instead?

Alternatively you could use `Sequence[ops.Gate]`, but those arguments should be converted to tuples on initialization to make them immutable.  This would let you also remove the `_as_sequence` conversion.

Reply:
`Union[ops.Gate, Tuple[ops.Gate, ...]]` is more like what I want since I want users to be assign just a gate to these fields.

>> those arguments should be converted to tuples on initialization

The class is defined as `dataclass(froze=True)` and `dataclasses` library doesn't have field transformers like (unlike `attrs`), unless I use `object.__setattr__` in `__post_init`, wdyt?

Reply:
I think it would be better to require either the (a) `tuple[ops.Gate, ...]` or (b) `Sequence[ops.Gate, ...]` type for the pre_q / post_q fields.  It is unlikely users would be defining these classes interactively; in such case it is OK to have them provide the sequence type that is stored in the instance.

For option (b) I am fine with using `object.__setattr__` in `__post_init` to convert a sequence argument to tuple.
If that feels ugly, I am also fine with adding a dependency on attrs and using attrs.frozen with a field converter here.

Converting to a tuple on initialization would let you remove the `_as_sequence` helper and property caching which effectively store gate sequences twice.

Reply:
used `attrs` with a `field(converter=)`